### PR TITLE
ams_grad optimizer implementation

### PR DESCRIPTION
**Pull request by haritha1313**
Reference to issue #1112.

### CODE REVIEW

**Comment by haritha1313** [comment]
@zoq the failed build log says-'_in "SimpleAmsGradTestFunction": absolute value of coordinates[1]{-0.31878442879857383} exceeds 0.10000000000000001_'. On testing locally, absolute value of all coordinates lie below 0.1. It would be great if you could help me understand what the trouble could be.

**Comment by rcurtin** [comment]
Your test is probably failing randomly due to the random seed being different on travis.  I'd suggest taking a look at the advice in this issue: https://github.com/mlpack/mlpack/issues/922 .

File: src/mlpack/core/optimizers/ams_grad/ams_grad.hpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+ * NumFunctions() should return the number of functions (\f$n\f$), and in the
+ * other two functions, the parameter i refers to which individual function (or
+ * gradient) is being evaluated.  So, for the case of a data-dependent function,
+ * such as NCA (see mlpack::nca::NCA), NumFunctions() should return the number
+ * of points in the dataset, and Evaluate(coordinates, 0) will evaluate the
+ * objective function on the first point in the dataset (presumably, the dataset
+ * is held internally in the DecomposableFunctionType).
+ */
+
+class AmsGrad
```
Comment (zoq) [inline src/mlpack/core/optimizers/ams_grad/ams_grad.hpp]: Do you think we could use AdamType as the base class and add another alias; as done for standard Adam and AdaMax: https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/optimizers/adam/adam.hpp#L167

File: src/mlpack/core/optimizers/ams_grad/amsgrad_update.hpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  /**
+   * Construct the AMSGrad update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient.
+   */
+  AmsGradUpdate(const double epsilon = 1e-8,
+             const double beta1 = 0.9,
```
Comment (zoq) [inline src/mlpack/core/optimizers/ams_grad/amsgrad_update.hpp]: Pedantic style issue, can you align the parameter (``beta1``, ``beta2``) with the first parameter (``epsilon``)?

File: src/mlpack/core/optimizers/ams_grad/amsgrad_update.hpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    m *= beta1;
+    m += (1 - beta1) * gradient;
+
+    v *= beta2;
+    v += (1 - beta2) * (gradient % gradient);
+
+    // Element wise maximum of past and present squared gradients.
+    vImproved = arma::max(vImproved, v);
+
+    iterate -= (stepSize * m) / (arma::sqrt(vImproved) + epsilon);
```
Comment (zoq) [inline src/mlpack/core/optimizers/ams_grad/amsgrad_update.hpp]: Note that the authors removed the debiasing step for simplicity, so the update should look like:

```
const double biasCorrection1 = 1.0 - std::pow(beta1, iteration);
const double biasCorrection2 = 1.0 - std::pow(beta2, iteration);

iterate -= (stepSize * std::sqrt(biasCorrection2) / biasCorrection1) *
    m / (arma::sqrt(vImproved) + epsilon);
```

File: src/mlpack/tests/ams_grad_test.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  AmsGrad optimizer(1e-3, 1, 0.9, 0.999, 1e-8, 500000, 1e-11, true);
+
+  arma::mat coordinates = f.GetInitialPoint();
+  optimizer.Optimize(f, coordinates);
+
+  BOOST_REQUIRE_SMALL(coordinates[0], 0.1);
+  BOOST_REQUIRE_SMALL(coordinates[1], 0.1);
+  BOOST_REQUIRE_SMALL(coordinates[2], 0.1);
+}
+
```
Comment (zoq) [inline src/mlpack/tests/ams_grad_test.cpp]: Peantic style issue number two, can you remove the extra line here?

File: src/mlpack/core/optimizers/ams_grad/ams_grad.hpp (hunk 1 above)
Comment (haritha1313) [reply src/mlpack/core/optimizers/ams_grad/ams_grad.hpp]: Yes, i agree with you. Will make the necessary changes.

File: src/mlpack/core/optimizers/adam/adam.hpp [hunk 5]
```diff
@@ -5,10 +5,10 @@
  * @author Marcus Edel
  * @author Vivek Pal
  *
- * Adam and AdaMax optimizer. Adam is an an algorithm for first-order gradient-
+ * Adam, AdaMax and AMSGrad optimizer. Adam is an an algorithm for first-order gradient-
```
Comment (zoq) [inline src/mlpack/core/optimizers/adam/adam.hpp]: Lines should be no more than 80 characters wide; for more information take a look at: https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping, the same applies to line 13 and 56.

Comment (haritha1313) [reply src/mlpack/core/optimizers/adam/adam.hpp]: I skipped that while editing. I have made edits to those lines.

File: src/mlpack/core/optimizers/adam/adam.hpp [hunk 6]
```diff
@@ -43,11 +45,15 @@ namespace optimization {
  *   year    = {2014},
  *   url     = {http://arxiv.org/abs/1412.6980}
  * }
+ * @article{
+ *   title   = {On the convergence of Adam and beyond},
+ *   url     = {https://openreview.net/pdf?id=ryQu7f-RZ}
+ * }
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/adam/adam.hpp]: We can at least add a year here if you like (2018 I guess), and then update this reference when the paper is published.

File: src/mlpack/core/optimizers/adam/amsgrad_update.hpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * AMSGrad is an exponential moving average variant which along with having 
+ * benefits of optimizers like Adam and RmsProp, also guarantees convergence.
+ * Unlike adam, it uses maximum of past squared gradients rather than their 
+ * exponential average for updation.
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/adam/amsgrad_update.hpp]: This is a really pedantic comment just for consistency---can you capitalize the same way the mlpack classes are?  So `adam` -> `Adam`, `RmsProp` -> `RMSProp`.

File: src/mlpack/core/optimizers/adam/amsgrad_update.hpp [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+ * For more information, see the following.
+ *
+ * @code
+ * @article{
+ *   title   = {On the convergence of Adam and beyond},
+ *   url     = {https://openreview.net/pdf?id=ryQu7f-RZ}
+ * }
+ * @endcode
+ */
+class AmsGradUpdate
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/adam/amsgrad_update.hpp]: Can we call this `AMSGradUpdate` to be consistent with the paper?  (The same applies for the `AmsGrad` typedef.)

**Comment by zoq** [comment]
Thanks again for the great contribution. You are now a contributor to mlpack, so if you like I can add your name and email in ``COPYRIGHT.txt`` and ``src/mlpack/core.hpp`` just let me know. Also, we can mail you some mlpack stickers if you want; just send me an email with your mailing address and we'll get them to you.

**Comment by haritha1313** [comment]
Sure! It was a great opportunity for me to learn, thanks for your reviews and help. And I do wish to contribute more to mlpack. Please let me know if you have any suggestions. And I will mail you the details.

**Pull request merged**